server/src/test/java/org/elasticsearch/search/vectors/AbstractKnnVectorQueryBuilderTestCase.java

```diff
@@ -0,0 +1,245 @@
+        TermQueryBuilder termQuery = new TermQueryBuilder("unmapped_field", 42);
+        KnnVectorQueryBuilder query = new KnnVectorQueryBuilder(VECTOR_FIELD, new float[] { 1.0f, 2.0f, 3.0f }, VECTOR_DIMENSION);
+        query.addFilterQuery(termQuery);
+
+        IllegalStateException e = expectThrows(IllegalStateException.class, () -> query.toQuery(context));
+        assertEquals("Rewrite first", e.getMessage());
+
+        QueryBuilder rewrittenQuery = query.rewrite(context);
+        assertThat(rewrittenQuery, instanceOf(MatchNoneQueryBuilder.class));
+    }
+
+    public void testOldVersionSerialization() throws IOException {
+        float[] bwcFloat = new float[VECTOR_DIMENSION];
+        KnnVectorQueryBuilder query = createTestQueryBuilder();
+        if (query.queryVector() != null) {
+            bwcFloat = query.queryVector();
+        } else {
+            for (int i = 0; i < query.getByteQueryVector().length; i++) {
+                bwcFloat[i] = query.getByteQueryVector()[i];
+            }
+        }
+        KnnVectorQueryBuilder queryWithNoFilters = new KnnVectorQueryBuilder(query.getFieldName(), bwcFloat, query.numCands()).queryName(
+            query.queryName()
+        ).boost(query.boost());
+
+        KnnVectorQueryBuilder queryNoByteQuery = new KnnVectorQueryBuilder(query.getFieldName(), bwcFloat, query.numCands()).queryName(
+            query.queryName()
+        ).boost(query.boost()).addFilterQueries(query.filterQueries());
+
+        Version newVersion = VersionUtils.randomVersionBetween(random(), Version.V_8_7_0, Version.CURRENT);
+        Version beforeByteQueryVersion = VersionUtils.randomVersionBetween(random(), Version.V_8_2_0, Version.V_8_6_0);
+        Version beforeFilterVersion = VersionUtils.randomVersionBetween(random(), Version.V_8_0_0, Version.V_8_1_0);
+
+        assertSerialization(query, newVersion);
+        assertSerialization(queryNoByteQuery, beforeByteQueryVersion);
+        assertSerialization(queryWithNoFilters, beforeFilterVersion);
+
+        for (var tuple : List.of(
+            Tuple.tuple(beforeByteQueryVersion, queryNoByteQuery),
+            Tuple.tuple(beforeFilterVersion, queryWithNoFilters)
+        )) {
+            try (BytesStreamOutput output = new BytesStreamOutput()) {
+                output.setVersion(tuple.v1());
+                output.writeNamedWriteable(query);
+                try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(), namedWriteableRegistry())) {
+                    in.setVersion(tuple.v1());
+                    KnnVectorQueryBuilder deserializedQuery = (KnnVectorQueryBuilder) in.readNamedWriteable(QueryBuilder.class);
+                    assertEquals(tuple.v2(), deserializedQuery);
+                    assertEquals(tuple.v2().hashCode(), deserializedQuery.hashCode());
+                }
+            }
+        }
+    }
+
+    @Override
+    public void testUnknownObjectException() throws IOException {
```

Comment:
nit: you could use assumeTrue(false) to signal that the tests were not executed, instead of being green although they did nothing.